File: python/tests/test_consumer_stats.py
# -*- coding: utf-8 -*-
"""消费侧统计（ConsumerStatsManager）单测 —— 离线，不依赖采样线程与集群。

Java 语义锚点（common/stats/StatsItem.java + client/stat/ConsumerStatsManager.java，5.5.1）：
* StatsItem 持有**累计** value/times；快照 = 累计差分（不是每分钟一个桶）；
  tps = sum*1000/(last.ts-first.ts)（**每秒**）；avgpt = sum/timesDiff（RT 项 = 平均耗时）；
* TPS 计数：addValue(key, msgs, 1) → tps = **调用次数**/秒，sum = 消息数；
* RT 计数：addRTValue(key, rt, 1) → avgpt = 平均毫秒；
* consumeStatus 全取 minute 快照；consumeFailedMsgs 取 failed 的 **hour** 窗口 sum。
"""
from __future__ import annotations

import time

from rocketmq.client.consumer_stats import (ConsumerStatsManager, StatsItem, StatsItemSet,
                                            compute_stats_data)
from rocketmq.remoting.protocol.body import ConsumeStatus

GROUP = "GID_StatsUnit"
TOPIC = "StatsTopic"


class TestComputeStatsData:
    def test_empty_list_gives_zero_snapshot(self):
        ss = compute_stats_data([])
        assert ss.sum == 0 and ss.tps == 0.0 and ss.avgpt == 0.0 and ss.times == 0

    def test_java_formula(self):
        # 两个累计点：10 秒内 value 增 30、times 增 3
        snaps = [(0, 100, 10), (10_000, 130, 13)]
        ss = compute_stats_data(snaps)
        assert ss.sum == 30
        assert abs(ss.tps - 3.0) < 1e-6          # 30 msgs / 10s = 3/s
        assert abs(ss.avgpt - 10.0) < 1e-6       # 30 / 3 = 10 per call
        assert ss.times == 3

    def test_single_sample_has_zero_span(self):
        # 只有一个点：span=0 → tps=0（Java 中除零防护来自 span>0 判断）
        ss = compute_stats_data([(1_000, 5, 1)])
        assert ss.sum == 0 and ss.tps == 0.0


class TestStatsItem:
    def test_cumulative_values(self):
        it = StatsItem("PULL_TPS", "T@G")
        it.add_value(5, 1)
        it.add_value(7, 1)
        assert it.value == 12
        assert it.times == 2

    def test_minute_snapshot_diffs_samples(self):
        it = StatsItem("PULL_TPS", "T@G")
        it.add_value(10, 1)
        it.sample()                    # 第一个累计点
        it.add_value(15, 1)
        it.sample()                    # 第二个累计点
        ss = it.get_stats_data_in_minute()
        assert ss.sum == 15
        assert ss.times == 1
        assert ss.tps >= 0

    def test_hour_window_separate_from_minute(self):
        # 累计差分：两点 (ts,100,2) → (ts,400,4)：sum=400-100=300、timesDiff=2
        it = StatsItem("PULL_RT", "T@G")
        it.add_value(100, 2)
        it.sample_hour()
        it.add_value(300, 2)
        it.sample_hour()
        ss = it.get_stats_data_in_hour()
        assert ss.sum == 300 and ss.times == 2
        assert abs(ss.avgpt - 150.0) < 1e-6

    def test_minute_list_capped(self):
        it = StatsItem("PULL_TPS", "T@G")
        for _ in range(80):
            it.sample()
        # 60 个点封顶（MINUTE_LIST_MAX）
        assert len(it._minute) == 60


class TestStatsItemSet:
    def test_get_and_create_is_idempotent(self):
        s = StatsItemSet("X")
        a = s.get_and_create("k")
        b = s.get_and_create("k")
        assert a is b

    def test_add_value_creates_and_accumulates(self):
        s = StatsItemSet("X")
        s.add_value("T@G", 3, 1)
        s.add_value("T@G", 4, 1)
        it = s.find("T@G")
        assert it is not None and it.value == 7 and it.times == 2
        assert s.find("other") is None


class TestConsumerStatsManager:
    def test_inc_semantics_match_java(self):
        # TPS 类：value=消息数、times=调用次数；RT 类：value=耗时、times=调用次数
        m = ConsumerStatsManager()
        m.inc_pull_tps(GROUP, TOPIC, 10)
        m.inc_pull_tps(GROUP, TOPIC, 20)
        item = m.topic_and_group_pull_tps.find("StatsTopic@GID_StatsUnit")
        assert item.value == 30 and item.times == 2

        m.inc_pull_rt(GROUP, TOPIC, 25)
        rt = m.topic_and_group_pull_rt.find("StatsTopic@GID_StatsUnit")
        assert rt.value == 25 and rt.times == 1

    def test_key_is_topic_at_group(self):
        m = ConsumerStatsManager()
        m.inc_consume_ok_tps("myGroup", "myTopic", 1)
        assert m.topic_and_group_consume_ok_tps.find("myTopic@myGroup") is not None

    def test_consume_status_all_zeros_when_untouched(self):
        m = ConsumerStatsManager()
        cs = m.consume_status(GROUP, TOPIC)
        assert isinstance(cs, ConsumeStatus)
        assert cs.pull_rt == 0 and cs.pull_tps == 0 and cs.consume_rt == 0
        assert cs.consume_ok_tps == 0 and cs.consume_failed_tps == 0
        assert cs.consume_failed_msgs == 0

    def test_consume_status_maps_rt_to_avgpt_and_tps_to_tps(self):
        # 直接喂累计快照点（间隔 1s），绕开真实采样线程，窗口内增量即窗口数据：
        #   pull_rt：value 增 200 / times 增 4 → avgpt=50ms
        #   consume_rt：value 增 40 / times 增 4 → avgpt=10ms
        #   pull_tps：value（消息数）增 20 → tps=20/s（Java 的 TPS 是**消息**每秒）
        #   consume_ok_tps：value 增 8 → tps=8/s
        m = ConsumerStatsManager()
        now = int(time.time() * 1000)

        def feed(s: StatsItemSet) -> None:
            it = s.find("StatsTopic@GID_StatsUnit")
            assert it is not None
            it._minute.clear()
            it._minute.append((now - 1000, 0, 0))
            it._minute.append((now, it.value, it.times))

        m.inc_pull_rt(GROUP, TOPIC, 50)
        m.inc_pull_rt(GROUP, TOPIC, 50)
        m.inc_pull_rt(GROUP, TOPIC, 50)
        m.inc_pull_rt(GROUP, TOPIC, 50)         # value=200 times=4
        for _ in range(4):
            m.inc_consume_rt(GROUP, TOPIC, 10)  # value=40 times=4
            m.inc_pull_tps(GROUP, TOPIC, 5)     # value=20 times=4
            m.inc_consume_ok_tps(GROUP, TOPIC, 2)  # value=8 times=4
        feed(m.topic_and_group_pull_rt)
        feed(m.topic_and_group_consume_rt)
        feed(m.topic_and_group_pull_tps)
        feed(m.topic_and_group_consume_ok_tps)
        cs = m.consume_status(GROUP, TOPIC)
        assert abs(cs.pull_rt - 50.0) < 1e-6
        assert abs(cs.consume_rt - 10.0) < 1e-6
        assert abs(cs.pull_tps - 20.0) < 1e-6
        assert abs(cs.consume_ok_tps - 8.0) < 1e-6

    def test_consume_failed_msgs_uses_hour_sum(self):
        m = ConsumerStatsManager()
        m.inc_consume_failed_tps(GROUP, TOPIC, 7)
        it = m.topic_and_group_consume_failed_tps.find("StatsTopic@GID_StatsUnit")
        now = int(time.time() * 1000)
        it._hour.append((now - 1000, 0, 0))
        it._hour.append((now, 7, 1))
        cs = m.consume_status(GROUP, TOPIC)
        assert cs.consume_failed_msgs == 7

    def test_start_shutdown_sampler(self):
        m = ConsumerStatsManager()
        m.start()
        try:
            assert m._thread is not None and m._thread.is_alive()
        finally:
            m.shutdown()
        assert m._thread is None

File: cpp/tests/test_consumer_stats.cpp
// ConsumerStatsManager + ConsumeStatus/ConsumerRunningInfo(307) 编码单测。
// 采样链不依赖真实时间窗（sum/tps 用窗口端点差分，测试直接喂采样点）。
#include <atomic>
#include <chrono>
#include <cstdio>
#include <string>
#include <thread>

#include "rocketmq/client/consumer_stats.h"
#include "rocketmq/common/util_all.h"

using namespace rocketmq;

namespace {
int g_checks = 0;
void expectEq(const std::string& got, const std::string& want, const char* what) {
    ++g_checks;
    if (got != want) {
        std::printf("FAIL %s: got [%s], want [%s]\n", what, got.c_str(), want.c_str());
        std::exit(1);
    }
}
void expectTrue(bool cond, const char* what) {
    ++g_checks;
    if (!cond) {
        std::printf("FAIL %s\n", what);
        std::exit(1);
    }
}
void expectNear(double got, double want, double eps, const char* what) {
    ++g_checks;
    if (got < want - eps || got > want + eps) {
        std::printf("FAIL %s: got %.6f, want %.6f\n", what, got, want);
        std::exit(1);
    }
}

std::string bodyStr(const Bytes& b) { return std::string(b.begin(), b.end()); }
}  // namespace

void testComputeStatsData() {
    // 空链全 0
    std::deque<std::tuple<int64_t, int64_t, int64_t>> empty;
    StatsSnapshot ss = computeStatsData(empty);
    expectEq(std::to_string(ss.sum), "0", "empty sum");
    expectEq(std::to_string(ss.times), "0", "empty times");

    // sum = 末点-首点；tps = sum*1000/span；avgpt = sum/timesDiff
    std::deque<std::tuple<int64_t, int64_t, int64_t>> chain = {
        {1000, 0, 0}, {2000, 300, 10}, {3000, 900, 20}};
    ss = computeStatsData(chain);
    expectEq(std::to_string(ss.sum), "900", "sum diff");
    expectNear(ss.tps, 450.0, 1e-9, "tps per second");
    expectEq(std::to_string(ss.times), "20", "times diff");
    expectNear(ss.avgpt, 45.0, 1e-9, "avgpt");
}

void testStatsItemSampling() {
    StatsItem item("PULL_RT", "T@G");
    item.addValue(100, 2);
    item.sample();
    item.addValue(200, 2);
    item.sample();
    StatsSnapshot m = item.getStatsDataInMinute();
    expectEq(std::to_string(m.sum), "200", "minute window sum (differential)");
    expectEq(std::to_string(m.times), "2", "minute window times");
    // 小时链是独立采样，小时窗口 sum 只到小时点
    item.sampleHour();
    StatsSnapshot h = item.getStatsDataInHour();
    expectEq(std::to_string(h.sum), "0", "hour chain has single point -> 0");
    expectEq(std::to_string(h.times), "0", "hour chain times");
}

void testManagerConsumeStatus() {
    ConsumerStatsManager m;
    // 直接注入相隔 1s 的两个采样点（与 Python 测试同款做法，不依赖真实时钟）
    const int64_t now = UtilAll::currentTimeMillis();
    auto feed = [&](StatsItemSet& set, int64_t v, int64_t t) {
        set.getAndCreate("StatsTopic@GID_StatsUnit")
            ->appendSampleForTest(now - 1000, 0, 0);
        set.getAndCreate("StatsTopic@GID_StatsUnit")
            ->appendSampleForTest(now, v, t);
    };
    feed(m.topicAndGroupPullRT(), 50, 1);
    feed(m.topicAndGroupPullTPS(), 5, 1);
    feed(m.topicAndGroupConsumeRT(), 10, 1);
    feed(m.topicAndGroupConsumeOKTPS(), 8, 1);
    feed(m.topicAndGroupConsumeFailedTPS(), 3, 1);

    ConsumeStatus cs = m.consumeStatus("GID_StatsUnit", "StatsTopic");
    expectNear(cs.pullRT, 50.0, 1e-6, "pullRT = avgpt");
    expectNear(cs.pullTPS, 5.0, 1e-6, "pullTPS = calls/sec");
    expectNear(cs.consumeRT, 10.0, 1e-6, "consumeRT = avgpt");
    expectNear(cs.consumeOKTPS, 8.0, 1e-6, "consumeOKTPS");
    expectNear(cs.consumeFailedTPS, 3.0, 1e-6, "consumeFailedTPS");
    // 未记录的 key 返回全 0
    ConsumeStatus zero = m.consumeStatus("GID_StatsUnit", "NoTopic");
    expectNear(zero.pullRT, 0.0, 1e-9, "unknown topic pullRT 0");
}

void testStartShutdown() {
    // 生命周期：start 后采样线程在跑，shutdown 干净退出（不挂死）
    ConsumerStatsManager m;
    m.start();
    m.incPullRT("G", "T", 1);
    std::this_thread::sleep_for(std::chrono::milliseconds(20));
    m.shutdown();
    m.shutdown();  // 幂等
    expectTrue(true, "start/shutdown lifecycle");
}

void testConsumeStatusJson() {
    ConsumeStatus cs;
    cs.pullRT = 1.5;
    cs.pullTPS = 2.0;
    cs.consumeRT = 3.0;
    cs.consumeOKTPS = 4.0;
    cs.consumeFailedTPS = 5.0;
    cs.consumeFailedMsgs = 6;
    JsonValue v = cs.toJson();
    ConsumeStatus back = ConsumeStatus::fromJson(v);
    expectNear(back.pullRT, 1.5, 1e-9, "consumeStatus json roundtrip pullRT");
    expectEq(std::to_string(back.consumeFailedMsgs), "6", "consumeFailedMsgs roundtrip");

    // ConsumerRunningInfo：statusTable/mqPopTable/userConsumerInfo 缺省输出空对象
    ConsumerRunningInfo ri;
    ri.properties[ConsumerRunningInfo::PROP_CONSUME_TYPE] = "CONSUME_PASSIVELY";
    ri.subscriptionSet = JsonValue::makeArray();
    const std::string s = bodyStr(ri.encode());
    expectTrue(s.find("\"statusTable\":{}") != std::string::npos, "statusTable empty object");
    expectTrue(s.find("\"mqPopTable\":{}") != std::string::npos, "mqPopTable empty object");
    expectTrue(s.find("\"userConsumerInfo\":{}") != std::string::npos, "userConsumerInfo empty object");
    // roundtrip
    ConsumerRunningInfo riBack;
    expectTrue(ConsumerRunningInfo::decode(ri.encode(), riBack), "decode ok");
    expectTrue(riBack.statusTable.isObject(), "statusTable roundtrip");
}

int main() {
    testComputeStatsData();
    testStatsItemSampling();
    testManagerConsumeStatus();
    testStartShutdown();
    testConsumeStatusJson();
    std::printf("consumer_stats: %d checks passed\n", g_checks);
    return 0;
}
